File: convServer.py
#!/usr/bin/env python
# Ahsen Uppal
# and
# Teo Georgiev

#
#  CS 6421 - Simple Python based conversion server. 
#  By default, it is an in to cm conversion server, but with optional
#  command-line arguments, it can run as any type of conversion
#  server.
#
#  Execution:    python convServer.py portnum [unit1 unit2]
#


import socket
import sys

length_units = {'ft' : 0.3048,
                'in' : 0.0254,
                'cm' : 0.01,
                'm'  : 1.0,
                'b'  : 0.1778,
}

mass_units = {'kg'  : 1.0,
              'g'   : 0.001,
              'lbs' : 0.45359237,
              'ounces' : 0.0283495231,
              'b'   : 0.125,
}

currency_units = {'$'  : 1.0,
                  'y'  : 0.008319,
                  'b'  : 0.19,
}

def process(conn, force_src_unit=None, force_dst_unit=None):
    greeting = "Welcome, you are connected to a Python-based %s to %s conversion server\n" % (force_src_unit, force_dst_unit)
    conn.send(greeting.encode('UTF-8'))
    userInput = conn.recv(BUFFER_SIZE).decode('UTF-8')
    if not userInput:
        print("Error reading message")
        return

    print("Received message: ", userInput)
    tokens = userInput.split()
    input_unit,output_unit,amount = tokens

    try:
        amount = float(amount)
    except:
        conn.send(("Invalid amount: %s\n" % (amount).encode('UTF-8')))
        return

    found = False
    for d in (length_units, mass_units, currency_units):
        if input_unit in d and output_unit in d:
            converted = amount * d[input_unit] / d[output_unit]
            found = True
            break
    if not found:
            conn.send(("Incompatible units %s and %s\n" % (input_unit,output_unit)).encode('UTF-8'))
            return

    if (force_src_unit and force_src_unit != input_unit) or (force_dst_unit and force_dst_unit != output_unit):
            conn.send(("Sorry, this is a %s to %s server\n" % (force_src_unit, force_dst_unit)).encode('UTF-8'))
            return

    conn.send(("%s\n" % converted).encode('UTF-8'))

if __name__ == '__main__':
    force_src_unit = 'in'
    force_dst_unit = 'cm'
    BUFFER_SIZE = 1024
    interface = ""

    if len(sys.argv) < 2:
        sys.stderr.write("usage: python {0} portnum\n".format(sys.argv[0]))
        sys.exit(1)

    portnum = int(sys.argv[1])

    if len(sys.argv) > 3:
        force_src_unit = sys.argv[2]
        force_dst_unit = sys.argv[3]

    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.bind((interface, portnum))
    s.listen(5)
    exit_flag = False
    try:
        print("Started Python-based %s to %s conversion server on port %s" % (force_src_unit, force_dst_unit, portnum))
        while not exit_flag:
            conn, addr = s.accept()
            print ('Accepted connection from client', addr)
            try:
                process(conn, force_src_unit, force_dst_unit)
            except:
                print ('Failed to process request from client.')
            conn.close()
    except KeyboardInterrupt:
        exit_flag = True

    print("Exiting...")
    s.close()
    sys.exit(0)

File: ConvServer.java
/*
 * Ahsen Uppal
 * and
 * Teo Georgiev
 */

/******************************************************************************
 *
 *  CS 6421 - Simple Conversation
 *  Compilation:  javac ConvServer.java
 *  Execution:    java ConvServer port [unit1 unit2 scale]
 *
 *  % java ConvServer portnum [unit1 unit2 scale]
 *
 * By default, this program runs a ft to in conversion server, but
 * with optional command-line arguments, it can run as any conversion
 * server.
 *
 ******************************************************************************/

import java.net.Socket;
import java.net.ServerSocket;
import java.io.IOException;
import java.io.PrintWriter;
import java.net.UnknownHostException;
import java.io.BufferedReader;
import java.io.InputStreamReader;

public class ConvServer {

    public static void process (Socket clientSocket, String unit1, String unit2, float scale) throws IOException {
        // open up IO streams
        BufferedReader in = new BufferedReader(new
InputStreamReader(clientSocket.getInputStream()));
        PrintWriter out = new PrintWriter(clientSocket.getOutputStream(),
true);

        /* Write a welcome message to the client */
        out.println("Welcome, you are connected to a Java-based server");

        /* read and print the client's request */
        // readLine() blocks until the server receives a new line from client
        String userInput;
        if ((userInput = in.readLine()) == null) {
            System.out.println("Error reading message");
            out.close();
            in.close();
            clientSocket.close();
        }

        System.out.println("Received message: " + userInput);

        String[] inputs = userInput.split(" ");
        if (inputs.length != 3)
        {
            System.out.println("Couldn't convert that - use format \'unit1 unit2 number\'");
        }

        double number = Float.parseFloat(inputs[2]);

        double answer;
        if (inputs[0].equals(unit1) && inputs[1].equals(unit2))
        {
            answer = number * scale;
            System.out.println(answer);
            out.println(answer);
        }
        else
        {
            String msg = "This server can only convert " + unit1 + " to " + unit2;
            System.out.println(msg);
            out.println(msg);
        }

        // close IO streams, then socket
        out.close();
        in.close();
        clientSocket.close();
    }

    public static void main(String[] args) throws Exception {
        String unit_src = "ft";
        String unit_dst = "in";
        float scale = 12;

        //check if argument length is invalid
        if (args.length < 1) {
            System.err.println("Usage: java ConvServer port [unit1 unit2 scale]");
        }

        // create socket
        int port = Integer.parseInt(args[0]);

        if (args.length > 3) {
            unit_src = args[1];
            unit_dst = args[2];
            scale = Float.parseFloat(args[3]);
        }

        ServerSocket serverSocket = new ServerSocket(port);
        System.err.println("Started Java-based " + unit_src + " to " + unit_dst + " conversion server on port " + port);

        // wait for connections, and process
        try {
            while(true) {
                // a "blocking" call which waits until a connection is requested
                Socket clientSocket = serverSocket.accept();
                System.err.println("\nAccepted connection from client");
                process(clientSocket, unit_src, unit_dst, scale);
            }

        }catch (IOException e) {
            System.err.println("Connection Error");
        }
        System.exit(0);
    }
}
